src/intersections/sweepline.jl

```diff
@@ -0,0 +1,240 @@
+          haskey(𝒞, new_geom) ? push!(𝒞[new_geom], _key(r), _key(below)) : (𝒞[new_geom] = [_key(r), _key(below)])
+        end
+      end
+      if u !== nothing
+        new_geom, new_type = _newevent(Segment(_key(u)), Segment(s))
+        if new_type == IntersectionType(0)
+          BinaryTrees.insert!(𝒬, new_geom)
+          haskey(𝒞, new_geom) ? push!(𝒞[new_geom], _key(u), s) : (𝒞[new_geom] = [_key(u), s])
+        end
+      end
+    end
+  end
+end
+
+_segdata(seg::Segment) = seg.vertices.data
+_key(node::BinaryTrees.AVLNode) = node.key
+_key(node::Nothing) = nothing
+_leftmost(node::BinaryTrees.AVLNode) = node.left === nothing ? node : _leftmost(node.left)
+_geom(intersect::Intersection) = intersect.geom
+_type(intersect::Intersection) = intersect.type
+function _newevent(s₁::Segment, s₂::Segment)
+  new_event = intersection(s₁, s₂)
+  if new_event !== nothing
+    _geom(new_event), _type(new_event)
+  else
+    nothing, nothing
+  end
+end
+
+# Helper: return the leftmost node (minimum) in a subtree.
+function _bst_minimum(node::BinaryTrees.AVLNode)
+  while node.left !== nothing
+    node = node.left
+  end
+  return node
+end
+
+# Helper: return the rightmost node (maximum) in a subtree.
+function _bst_maximum(node::BinaryTrees.AVLNode)
+  while node.right !== nothing
+    node = node.right
+  end
+  return node
+end
+
+"""
+    find_above_below(root, x)
+
+Find the node above and below `x` in the binary search tree rooted at `root`.
+Returns a tuple `(above, below)` where `above` is the node with the smallest key
+greater than `x.key` and `below` is the node with the largest key smaller than `x.key`.
+If `x` is not found, returns the best candidates for `above` and `below`.
+"""
+function find_above_below(root::BinaryTrees.AVLNode, x::BinaryTrees.AVLNode)
+  above = nothing
+  below = nothing
+  current = root
+  # Traverse from the root to the target node, updating candidates.
+  while current !== nothing && current.key != x.key
+    if x.key < current.key
+      # current is a potential above (successor)
+      above = current
+      current = current.left
+    else # x.key > current.key
+      # current is a potential below (predecessor)
+      below = current
+      current = current.right
+    end
+  end
+
+  # If the node wasn't found, return the best candidate values
+  if current === nothing
+    return (above, below)
+  end
+
+  # Found the node with key equal to x.key.
+  # Now, if there is a left subtree, the true below (predecessor) is the maximum in that subtree.
+  if current.left !== nothing
+    below = _bst_maximum(current.left)
+  end
+  # Similarly, if there is a right subtree, the true above (successor) is the minimum in that subtree.
+  if current.right !== nothing
+    above = _bst_minimum(current.right)
+  end
+
+  (above, below)
+end
+find_above_below(root::BinaryTrees.AVLNode, x::Nothing) = (nothing, nothing)
+function Segment(node₁::BinaryTrees.BinaryNode, node₂::BinaryTrees.BinaryNode)
```

Comment:
Please avoid introducing a function with the same name of an existing geometry type. That can generate invalidations with the Julia compiler, and can be a source of potential confusion.